### ISSUE TITLE
Modules use EC2 role instead pod role to fetch the data from S3 

### ISSUE THREAD

**Issue by sdwerwed** [hashicorp/terraform#28101]
**Terraform version:**
`terraform_0.15.0-beta1`
`terraform_0.14.8`
`terraform_0.13.0-beta2`

We use a module from s3 bucket from other AWS account that requires authentication to get the module. 

**This **don't work** in EKS when environment variables are set:**
```
AWS_ROLE_ARN=arn:aws:iam::xxxxxxxxxxxxxx:role/atlantis
AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

**This works when env variables are set manually: (old)**
```
AWS_ACCESS_KEY_ID=ASIAxxxxxxxxxxxxx
AWS_SECRET_ACCESS_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
AWS_SESSION_TOKEN=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx..
```

**How to fix it: (old)**
Could you please use the `AWS_ACCESS_KEY_ID` `AWS_SECRET_ACCESS_KEY` `AWS_SESSION_TOKEN` generated by
`aws sts assume-role-with-web-identity --role-arn arn:aws:iam::xxxxxxxx:role/atlantis --role-session-name app1 --web-identity-token cat $AWS_WEB_IDENTITY_TOKEN_FILE` and then try to fetch the modules by providing those secrets on api?

**Comment by sdwerwed** [comment]
`terraform init -reconfigure` fixed it, without creating the temporary credentials. It was needed for the first time, I am leaving it open to see if it was expected or if it is a bug.

**Comment by sdwerwed** [comment]
After more research, the module is trying to download with EC2 role instead of pod role.  aws cli works but modules not.

I tried also this but it did not work:
```
provider "aws" {
  region = "eu-west-1"
  shared_credentials_file = "/root/.aws/config"
  profile = "default"
  skip_metadata_api_check = true

}
```

**Comment by amine250** [comment]
I'm running a terraform module with an S3 backend inside a pod in EKS. I'm having the same issue with the latest Terraform (0.14.8).

By default, I noticed that Terraform automatically uses the EC2 role instead of the pod role, even when `AWS_ROLE_ARN` and `AWS_WEB_IDENTITY_TOKEN_FILE` are set. And when using `skip_metadata_api_check = true`, I get this when running `terraform init` : 
```sh
Initializing the backend...

Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.

Please see https://www.terraform.io/docs/language/settings/backends/s3.html
for more information about providing credentials.

Error: NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

**Comment by sdwerwed** [comment]
@pselle I think this issue should stay open for now based on that https://github.com/hashicorp/terraform-provider-aws/issues/18140#issuecomment-801427235

**Comment by pselle** [comment]
@sdwerwed Reopened, thank you for commenting!

**Comment by crw** [comment]
Thanks for that comment. I will close this as a duplicate of #26292. If I am missing a nuance that makes these separate issues, please let me know.

**Comment by md5** [comment]
#26292 was already closed duplicate of this issue. One of them should be kept open since this is a real problem that has not been fixed

**Comment by crw** [comment]
Somehow I completely missed that. Sorry!

**Comment by md5** [comment]
No problem. Should the `duplicate` label be removed?

**Comment by Anton-Sagurov** [comment]
Hello, just faced the same issue. 
I just want to let you know - this is still actual.

**Comment by gdavison** [comment]
This will be addressed in Terraform v1.6. If you still see this error when v1.6 is released (or you try any of the v1.6 beta releases), please open a new issue
